File: packages/flutter-im-ui/test/view_state_vectors_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flare_im_ui/flare_im_ui.dart';
import 'package:flutter_test/flutter_test.dart';

// FR-057: a failed refresh over rows worth keeping does not wipe what someone was reading.

void main() {
  test('presentation matches the shared table', () {
    final table =
        jsonDecode(
              File(
                '${Directory.current.path}/../../spec/view-state-vectors.json',
              ).readAsStringSync(),
            )
            as Map<String, dynamic>;
    final cases = (table['cases'] as List).cast<Map<String, dynamic>>();
    expect(cases.length, greaterThanOrEqualTo(10));
    for (final c in cases) {
      final status = FlareViewStatus.values.firstWhere(
        (s) => s.name == c['status'] as String,
      );
      expect(
        flareViewPresentation(status, c['stale'] as bool).name,
        c['presentation'] as String,
        reason: c['id'] as String,
      );
    }
  });
}

File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/ViewStateVectorsTest.kt
package com.flare.im.ui

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/** FR-057: a failed refresh over rows worth keeping does not wipe what someone was reading. */
class ViewStateVectorsTest {
    private fun vectorsFile(): File {
        var dir: File? = File(".").absoluteFile
        while (dir != null) {
            val candidate = File(dir, "spec/view-state-vectors.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile
        }
        error("spec/view-state-vectors.json not found")
    }

    @Test fun presentationMatchesTheSharedTable() {
        val cases = (FlareJson.parse(vectorsFile().readText()) as Map<*, *>)["cases"] as List<*>
        assertTrue(cases.size >= 10, "the shared table lost cases")
        for (raw in cases) {
            val case = raw as Map<*, *>
            val status = FlareApplicationViewStatus.entries.first { it.name.equals(case["status"] as String, ignoreCase = true) }
            val presentation = flareViewPresentation(status, case["stale"] as Boolean)
            assertEquals(case["presentation"], presentation.name.replaceFirstChar { it.lowercase() }, case["id"] as String)
        }
    }
}
